### DISCUSSION TITLE
Preserving Orientation when Switching Particle Types in Overdamped BD

### DISCUSSION THREAD

**Discussion by idop11** [glotzerlab/hoomd-blue#1975] [Support]
Hello everyone,

I’m running a 2D simulation in HOOMD‐blue where the simulation box is conceptually split into two regions along the y-axis:
- The **top** region $(y>0)$ contains `type A` particles that have a nonzero active force.
- The **bottom** region $(y\leq0)$ contains `type B` particles that have zero active force.

As soon as a particle crosses from $(y>0)$ into $(y\leq0)$ (or vice versa), I change its `typeid` accordingly. In principle, I want all particles to maintain a continuous orientation and to undergo rotational diffusion with time scale $\tau_r = 1/D_r$. However, in practice, **once a particle switches type, it doesn’t appear to keep the orientation it had before the switch** - it looks like the orientation resets or changes abruptly.

Here’s a minimal reproducible example that illustrates my approach. The essential points are:
1. I use a custom `TypeSwitchAction` to change `typeid` based on a particle’s $y$-position.
2. I have two forces:
- `active `(filter=type A) with a nonzero active force,
- `activeB` (filter=type B) with zero active force.
3. I apply `ActiveRotationalDiffusion` updaters to **both** active forces so that all particles should keep rotating.

Despite this setup, particles sometimes fail to preserve their pre‐switch orientation. I’d greatly appreciate any pointers or insights from folks who have encountered similar behavior or have ideas on how to fix this.

My script:
[sample_code.txt](https://github.com/user-attachments/files/18240092/sample_code.txt)

**Problem**:
I need each particle’s orientation quaternion to remain continuous (and keep rotating stochastically) when crossing from $(y>0)$ to $(y\leq0)$, but it appears that the orientation is reset or changed abruptly at the type switch.
Any help identifying the cause of this reset (and how to preserve orientation seamlessly) would be greatly appreciated. 

Thanks in advance!

**Comment by joaander** [comment]
The only obvious problem I see in your code is that the filters are not updated. `type_C` and `type_B` will continue to refer to the original particles that matched the filter even though you change the type. Add a `FilterUpdater` to reevaluate the filters: https://hoomd-blue.readthedocs.io/en/v5.0.0/hoomd/update/filterupdater.html

Could you provide a smaller minimum example? Ideally you would place only one particle in the simulation box and set the initial position/orientation such that running a few timesteps is enough to trigger the behavior you observe. No GSD file is necessary, `take_snapshot` and `print()` should be sufficient to demonstrate that orientation is reset.
